rumqttd/src/link/bridge.rs

```diff
@@ -198,62 +198,50 @@ pub async fn tls_connect<P: AsRef<Path>>(
 ) -> Result<Box<dyn N>, BridgeError> {
     let mut root_cert_store = RootCertStore::empty();
 
-    let ca_certs = rustls_pemfile::certs(&mut BufReader::new(Cursor::new(fs::read(ca_file)?)))?;
-    let trust_anchors = ca_certs.iter().map_while(|cert| {
-        if let Ok(ta) = webpki::TrustAnchor::try_from_cert_der(&cert[..]) {
-            Some(OwnedTrustAnchor::from_subject_spki_name_constraints(
-                ta.subject,
-                ta.spki,
-                ta.name_constraints,
-            ))
-        } else {
-            None
-        }
-    });
-
-    root_cert_store.add_trust_anchors(trust_anchors);
+    let ca_certs = rustls_pemfile::certs(&mut BufReader::new(Cursor::new(fs::read(ca_file)?)))
+        .collect::<Result<Vec<_>, _>>()?;
+    root_cert_store.add_parsable_certificates(ca_certs);
 
     if root_cert_store.is_empty() {
         return Err(BridgeError::NoValidCertInChain);
     }
 
-    let config = ClientConfig::builder()
-        .with_safe_defaults()
-        .with_root_certificates(root_cert_store);
+    let config = ClientConfig::builder().with_root_certificates(root_cert_store);
 
     let config = if let Some(ClientAuth {
         certs: certs_path,
         key: key_path,
     }) = client_auth_opt
     {
         let read_certs =
-            rustls_pemfile::certs(&mut BufReader::new(Cursor::new(fs::read(certs_path)?)))?;
+            rustls_pemfile::certs(&mut BufReader::new(Cursor::new(fs::read(certs_path)?)))
+                .collect::<Result<Vec<_>, _>>()?;
 
-        let read_keys = match rustls_pemfile::read_one(&mut BufReader::new(Cursor::new(fs::read(
+        let read_key = match rustls_pemfile::read_one(&mut BufReader::new(Cursor::new(fs::read(
             key_path,
         )?)))? {
-            Some(rustls_pemfile::Item::RSAKey(_)) => rustls_pemfile::rsa_private_keys(
+            Some(rustls_pemfile::Item::Pkcs1Key(_)) => rustls_pemfile::rsa_private_keys(
                 &mut BufReader::new(Cursor::new(fs::read(key_path)?)),
-            )?,
-            Some(rustls_pemfile::Item::PKCS8Key(_)) => rustls_pemfile::pkcs8_private_keys(
+            )
+            .next()
+            .ok_or(BridgeError::NoValidCertInChain)??
+            .into(),
+            Some(rustls_pemfile::Item::Pkcs8Key(_)) => rustls_pemfile::pkcs8_private_keys(
                 &mut BufReader::new(Cursor::new(fs::read(key_path)?)),
-            )?,
+            )
+            .next()
+            .ok_or(BridgeError::NoValidCertInChain)??
+            .into(),
```

Comment:
why do we need to re-read the same key? this would be same as:

```rust
let read_key = match rustls_pemfile::read_one(&mut BufReader::new(Cursor::new(fs::read(
    key_path,
)?)))? {
    Some(rustls_pemfile::Item::Pkcs1Key(key)) => key.into(),
    Some(rustls_pemfile::Item::Pkcs8Key(key)) => key.into(),
    Some(rustls_pemfile::Item::Sec1Key(key)) => key.into(),
    None | Some(_) => return Err(BridgeError::NoValidCertInChain),
};

```
also, i think `read_one` shall be called in loop ( as we should ignore _other_ `Item` which aren't _keys_, just like `xx_private_keys()` fn do ), so:

```rust
let read_key = loop {
    match rustls_pemfile::read_one(&mut BufReader::new(Cursor::new(fs::read(key_path)?)))? {
        Some(rustls_pemfile::Item::Pkcs1Key(key)) => break key.into(),
        Some(rustls_pemfile::Item::Pkcs8Key(key)) => break key.into(),
        Some(rustls_pemfile::Item::Sec1Key(key)) => break key.into(),
        Some(_) => {}
        None => return Err(BridgeError::NoValidCertInChain),
    }
};

```

Reply:
Good question, the existing code was doing it this way so I assumed there was some reason for it.

The existing code doesn't support sec1 or ignoring non-key items so maybe let's change that in a separate PR, rather than mixing behaviour changes with dependency updates.

Reply:
> The existing code doesn't support sec1 or ignoring non-key items so maybe let's change that in a separate PR, rather than mixing behaviour changes with dependency updates.
>

fair point, lets open PR after this PR is merged :)

Reply:
hey @qwandor , would you like to open PR for this?

Reply:
I have opened the PR: https://github.com/bytebeamio/rumqtt/pull/802 :)